src/MatrixProductOperator.jl

```diff
@@ -0,0 +1,175 @@
+using OptimizedEinsum: ContractionPath
+using UUIDs: uuid4
+using IterTools: partition
+using Random
+
+abstract type MatrixProductOperator{B} <: State{B} end
+
+function MatrixProductOperator(arrays; bounds::Type{<:Bounds} = Open, kwargs...)
+    MatrixProductOperator{bounds}(arrays; kwargs...)
+end
+
+function MatrixProductOperator{Open}(arrays; χ = nothing, order = (:l, :r, :p), meta...)
+    !issetequal(order, (:l, :r, :p)) && throw(ArgumentError("`order` must be a permutation of the :l, :r and :p"))
+    order = Dict(dim => i for (i, dim) in enumerate(order))
+
+    # check format
+    # NOTE matching index dimensions are checked on `push!(::TensorNetwork,...)`
+    # TODO check χ?
+    if !all(==(4) ∘ ndims, arrays[2:end-1]) || ndims(first(arrays)) != 3 || ndims(last(arrays)) != 3
+        throw(DimensionMismatch("virtual bond-dims mismatch"))
+    end
+
+    tn = TensorNetwork{MatrixProductOperator{Open}}(; χ = χ, order = order, meta...)
+
+    n = length(arrays)
+    vinds = Dict(x => Symbol(uuid4()) for x in partition(1:n, 2, 1))
+    oinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    iinds = Dict(i => Symbol(uuid4()) for i in 1:n)
+    permutator = [order[i] for i in (:l, :r, :p)]
+
+    # add boundary tensors
+    # first
+    labels = if order[:r] < order[:p]
+        (vinds[(1, 2)], iinds[1], oinds[1])
+    else
+        (iinds[1], oinds[1], vinds[(1, 2)])
+    end
+    push!(tn, Tensor(first(arrays), labels))
+    # last
+    labels = if order[:l] < order[:p]
+        (vinds[(n - 1, n)], iinds[n], oinds[n])
+    else
+        (iinds[n], oinds[n], vinds[(n - 1, n)])
+    end
+    push!(tn, Tensor(last(arrays), labels))
+
+    # add other tensors
+    for (i, data) in zip(2:n-1, arrays[2:end-1])
+        lind = vinds[(mod1(i - 1, n), i)]
+        rind = vinds[(i, mod1(i + 1, n))]
+
+        labels = [lind, rind, iinds[i], oinds[i]]
+        permute!(labels, permutator)
+
+        tensor = Tensor(data, labels)
+        push!(tn, tensor)
+    end
+
+    # mark input indices
+    for (site, label) in iinds
+        index = inds(tn, label)
+        index.meta[:site] = site
+        index.meta[:plug] = :input
+    end
+    # mark output indices
+    for (site, label) in oinds
+        index = inds(tn, label)
+        index.meta[:site] = site
+        index.meta[:plug] = :output
+    end
+
+    return tn
+end
+
+function MatrixProductOperator{Closed}(arrays; χ = nothing, order = (:l, :r, :p), meta...)
+    !issetequal(order, (:l, :r, :p)) && throw(ArgumentError("`order` must be a permutation of the :l, :r and :p"))
+    order = Dict(side => i for (i, side) in enumerate(order))
```

Comment:
Same as above, you are not checking for input/output indices.

Reply:
Thanks!